vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -210,97 +183,177 @@ type internal FSharpGoToDefinitionService [<ImportingConstructor>]
+                    else
+                        let! sigDocument = originDocument.Project.Solution.TryGetDocumentFromPath targetRange.FileName
+                        let! sigSourceText = sigDocument.GetTextAsync () |> liftTaskAsync
+                        let sigTextSpan = CommonRoslynHelpers.FSharpRangeToTextSpan (sigSourceText, targetRange)
+                        // if gotodef call originated in a signature and the returned target is a signature, navigate to there
+                        if isSignatureFile targetRange.FileName && preferSignature then 
+                            let navItem = FSharpNavigableItem (sigDocument, sigTextSpan)
+                            results.Add navItem
+                            return results.AsEnumerable()
+                        else // we need to get an FSharpSymbol from the targetRange found in the signature
+                                // that symbol will be used to find the destination in the corresponding implementation file
+                            let implFilePath = Path.ChangeExtension (sigDocument.FilePath,"fs")
+                            let! implDocument = originDocument.Project.Solution.TryGetDocumentFromPath implFilePath
+                            let! implVersion = implDocument.GetTextVersionAsync () |> liftTaskAsync
+                            let! implSourceText = implDocument.GetTextAsync () |> liftTaskAsync
+                            let! projectOptions = projectInfoManager.TryGetOptionsForEditingDocumentOrProject implDocument
+                            let! targetRange = 
+                                findSymbolDeclarationInFile 
+                                    (targetSymbolUse, implFilePath, implSourceText.ToString(), checkerProvider.Checker, projectOptions, implVersion.GetHashCode())
+                                    |> Async.RunSynchronously
+                            let implTextSpan = CommonRoslynHelpers.FSharpRangeToTextSpan (implSourceText, targetRange)
+                            let navItem = FSharpNavigableItem (implDocument, implTextSpan)
+                            results.Add navItem
+                            return results.AsEnumerable()
+                    | _ -> return! None
+            with e ->
+                debug "\n%s\n%s\n%s\n" e.Message (e.TargetSite.ToString()) e.StackTrace
+                return Seq.empty
+        }   |> Async.map (Option.defaultValue Seq.empty)
+            |> CommonRoslynHelpers.StartAsyncAsTask cancellationToken        
 
-        
+   
     interface IGoToDefinitionService with
         
-        member this.FindDefinitionsAsync (document: Document, position: int, cancellationToken: CancellationToken) =
-            this.FindDefinitionsTask (document, position, cancellationToken)
+        member this.FindDefinitionsAsync(document: Document, position: int, cancellationToken: CancellationToken) =
+            this.FindDefinitionsTask(document, position, cancellationToken)
```

Comment:
so that the function can be tested using a stripped down workspace without needing to rely on MEF imports or mocks